### PR TITLE
[Java] Allow configuration of OkHttpClient.Builder

### LINKED ISSUE #15404: Support certificate pinning for java SignalR client
### Is your feature request related to a problem? Please describe.
For an android application I need to pin the server TLS certificate of the SignalR connection.

The [old SignalR library](https://github.com/SignalR/java-client) had the possibility to provide a custom HttpConnection implementation (via ServerSentEventsTransport). We used that to provide our OkHttpClient instance, that did the certificate pinning.

As far as I could see, the new java library for the .net core SignalR doesn't provide a way to pin the certificate.
There is a `withHttpClient` method in the [Builder](https://github.com/aspnet/AspNetCore/blob/3d3ad96206fc504dba954c0c6403139e3438263e/src/SignalR/clients/java/signalr/src/main/java/com/microsoft/signalr/HttpHubConnectionBuilder.java), that could probably be used to provide an http client with the pinning, but that method and the corresponding `HttpClient` class are package private, so we can't use them.

### Describe the solution you'd like
The most flexible solution would probably be to expose the `withHttpClient` method an `HttpClient` abstract class, so that the library users can provide custom implementations.

If there is some other way to implement certificate pinning or if I overlooked something, I look forward for feedback.

### PR DESCRIPTION
**Pull request by BrennanConroy**
Fixes https://github.com/dotnet/aspnetcore/issues/15404

No functional tests in java currently. Manually tested out setting a cert pinner and read timeout on the builder, looks good.

New method on the `HttpHubConnectionBuilder` class.
`HttpHubConnectionBuilder setHttpClientBuilderCallback(Action1<OkHttpClient.Builder> configureBuilder)`

### CODE REVIEW

File: src/SignalR/clients/java/signalr/src/test/java/com/microsoft/signalr/WebSocketTransportTest.java [hunk 1]
```diff
@@ -17,11 +17,11 @@
...
-    public void WebSocketThrowsIfItCantConnect() {
-        Transport transport = new WebSocketTransport(new HashMap<>(), new DefaultHttpClient());
-        RuntimeException exception = assertThrows(RuntimeException.class, () -> transport.start("http://url.fake.example").blockingAwait(1, TimeUnit.SECONDS));
-        assertEquals("There was an error starting the WebSocket transport.", exception.getMessage());
-    }
+    // public void WebSocketThrowsIfItCantConnect() {
+    //     Transport transport = new WebSocketTransport(new HashMap<>(), new DefaultHttpClient());
+    //     RuntimeException exception = assertThrows(RuntimeException.class, () -> transport.start("http://url.fake.example").blockingAwait(1, TimeUnit.SECONDS));
+    //     assertEquals("There was an error starting the WebSocket transport.", exception.getMessage());
+    // }
```
Comment (halter73) [inline src/SignalR/clients/java/signalr/src/test/java/com/microsoft/signalr/WebSocketTransportTest.java:24]: Is commenting out just the `@Test` attribute no longer sufficient?

File: src/SignalR/clients/java/signalr/src/main/java/com/microsoft/signalr/HttpHubConnectionBuilder.java [hunk 2]
```diff
@@ -113,12 +115,25 @@ public HttpHubConnectionBuilder withHeader(String name, String value) {
         return this;
     }
 
+    /**
+     * Adds a method that will be called when constructing the HttpClient to allow customization such as certificate validation, proxies, and cookies.
```
Comment (halter73) [inline src/SignalR/clients/java/signalr/src/main/java/com/microsoft/signalr/HttpHubConnectionBuilder.java:119]: We should probably call out that calling configureHttpClient multiple times overrides the previous configureBuilder.

File: src/SignalR/clients/java/signalr/src/test/java/com/microsoft/signalr/WebSocketTransportTest.java (hunk 1 above)
Comment (BrennanConroy) [reply src/SignalR/clients/java/signalr/src/test/java/com/microsoft/signalr/WebSocketTransportTest.java:24]: I removed the DefaultHttpClient constructor that was being used. Maybe I should just delete the test

Comment (halter73) [reply src/SignalR/clients/java/signalr/src/test/java/com/microsoft/signalr/WebSocketTransportTest.java:24]: That or just call the new constructor with a no-op builder. Up to you.

File: src/SignalR/clients/java/signalr/src/main/java/com/microsoft/signalr/HttpHubConnectionBuilder.java (hunk 2 above)
Comment (davidfowl) [reply src/SignalR/clients/java/signalr/src/main/java/com/microsoft/signalr/HttpHubConnectionBuilder.java:119]: Maybe it should run all

Comment (JamesNK) [reply src/SignalR/clients/java/signalr/src/main/java/com/microsoft/signalr/HttpHubConnectionBuilder.java:119]: If the last set wins then the method name should be `setConfigureHttpClient(...)`. The comment would be "~Adds~ Sets a method that"

**Comment by BrennanConroy** [comment]
Any other feedback here?

File: src/SignalR/clients/java/signalr/src/main/java/com/microsoft/signalr/HttpHubConnectionBuilder.java [hunk 3]
```diff
@@ -113,12 +115,25 @@ public HttpHubConnectionBuilder withHeader(String name, String value) {
...
     }
 
+    /**
+     * Sets a method that will be called when constructing the HttpClient to allow customization such as certificate validation, proxies, and cookies.
+     * By default the client will have a cookie jar added and a read timeout for LongPolling.
+     *
+     * @param configureBuilder Callback for configuring the OkHttpClient.Builder.
+     * @return This instance of the HttpHubConnectionBuilder.
+     */
+    public HttpHubConnectionBuilder setConfigureHttpClient(Action1<OkHttpClient.Builder> configureBuilder) {
```
Comment (halter73) [inline src/SignalR/clients/java/signalr/src/main/java/com/microsoft/signalr/HttpHubConnectionBuilder.java:125]: I liked using `set` to make it clear this is effectively property setter, but seeing it paired with the verb "configure" is making me hesitant. Generally, I'd want to be setting a noun.

For example, I think `setHttpClientConfigurationCallback` sounds better despite maybe being a bit too wordy.

Does anyone have any other suggestions? Am I the only one who cares about the noun vs verb thing?

Comment (BrennanConroy) [reply src/SignalR/clients/java/signalr/src/main/java/com/microsoft/signalr/HttpHubConnectionBuilder.java:125]: > Am I the only one who cares about the noun vs verb thing?

You're not alone.

`setHttpClientBuilderCallback`?

Comment (BrennanConroy) [reply src/SignalR/clients/java/signalr/src/main/java/com/microsoft/signalr/HttpHubConnectionBuilder.java:125]:
```suggestion
    public HttpHubConnectionBuilder setHttpClientBuilderCallback(Action1<OkHttpClient.Builder> configureBuilder) {
```

File: src/SignalR/clients/java/signalr/src/main/java/com/microsoft/signalr/DefaultHttpClient.java [hunk 4]
```diff
@@ -19,14 +19,14 @@
...
-        this(0, null);
+    public DefaultHttpClient(Action1<OkHttpClient.Builder> configureBuilder) {
+        this(0, null, configureBuilder);
     }
 
     public DefaultHttpClient cloneWithTimeOut(int timeoutInMilliseconds) {
         OkHttpClient newClient = client.newBuilder().readTimeout(timeoutInMilliseconds, TimeUnit.MILLISECONDS)
                 .build();
-        return new DefaultHttpClient(timeoutInMilliseconds, newClient);
+        return new DefaultHttpClient(timeoutInMilliseconds, newClient, null);
```
Comment (halter73) [inline src/SignalR/clients/java/signalr/src/main/java/com/microsoft/signalr/DefaultHttpClient.java:29]: Should we store the configureBuilder and reapply it for cloned clients? Long term, we should add timeouts with something other than cloneWithTimeOut, but for now, it seems broken not to.

Comment (BrennanConroy) [reply src/SignalR/clients/java/signalr/src/main/java/com/microsoft/signalr/DefaultHttpClient.java:29]: The cloned client gets the same configuration applied. It's using the same builder internally effectively.

Comment (halter73) [reply src/SignalR/clients/java/signalr/src/main/java/com/microsoft/signalr/DefaultHttpClient.java:29]: Wait a minute... It looks like `timeoutInMilliseconds` is [completely ignored](https://github.com/dotnet/aspnetcore/blob/b5690db723351504fe96785b0bbebd0b0699f757/src/SignalR/clients/java/signalr/src/main/java/com/microsoft/signalr/DefaultHttpClient.java#L41) when a non-null client is passed in. I guess there's no need to pass in `configureBuilder` if it wouldn't be called anyway, but this looks completely broken to me. How do timeouts work if timeoutInMilliseconds is just being thrown away? Did it ever work? Am I missing something obvious?

Comment (BrennanConroy) [reply src/SignalR/clients/java/signalr/src/main/java/com/microsoft/signalr/DefaultHttpClient.java:29]: The constructors are a bit odd here, could use some cleanup. The timeout is actually used 2 lines above and unused when passed in here.

Comment (halter73) [reply src/SignalR/clients/java/signalr/src/main/java/com/microsoft/signalr/DefaultHttpClient.java:29]: We're going to reapply the configureBuilder then, right?

Comment (BrennanConroy) [reply src/SignalR/clients/java/signalr/src/main/java/com/microsoft/signalr/DefaultHttpClient.java:29]: `newBuilder` already copies the settings that were applied with the `configureBuilder`.

File: src/SignalR/clients/java/signalr/src/main/java/com/microsoft/signalr/DefaultHttpClient.java [hunk 5]
```diff
@@ -36,7 +36,7 @@ public void close() {
         }
     }
 
-    public DefaultHttpClient(int timeoutInMilliseconds, OkHttpClient client) {
+    public DefaultHttpClient(int timeoutInMilliseconds, OkHttpClient client, Action1<OkHttpClient.Builder> configureBuilder) {
```
Comment (halter73) [inline src/SignalR/clients/java/signalr/src/main/java/com/microsoft/signalr/DefaultHttpClient.java:39]: Can we remove the `timeoutInMilliseconds` parameter then? I don't see anywhere it's passed in with a null client, and with a non-null client it doesn't do anything.

**Review by halter73 — APPROVED** [review]

**Comment by BrennanConroy** [comment]
@dotnet/aspnet-build could I get a merge please? I already talked to Kevin
